### ISSUE TITLE
Crash during child/IPsec SA rekey

### ISSUE THREAD

**Issue by kavinda-forcepoint** [libreswan/libreswan#505]
### Summary

In Libreswan 4.5 _with_ labeled IPsec, the Gateway triggered the following assertion during child/IPsec SA re-keying:

```
ABORT: ASSERTION FAILED: *chosen_proposal == ((void *)0) (ikev2_process_sa_payload() +1138 programs/pluto/ikev2_spdb_struct.c)
```

### Details

* `*chosen_proposal` in the assert message refers to `child->sa.st_accepted_esp_or_ah_proposal`.
  * Let's call this `child`, `crashChild`.
* Serial number of `crashChild` = 243520.
  * Let's call this `crashChildSerial`.
* Serial number of the `ike_sa` associated with `crashChild` _at the time of crash_ = 243550
  * Let's call this IKE, `crashIKE`.
  * Let's call this serial `crashIKESerial`.
* `crashChild.st_accepted_esp_or_ah_proposal` was initially set 1 minute before the crash.
* Serial number of the `ike_sa` associated with `crashChild` _at the time of initially setting_ `st_accepted_esp_or_ah_proposal` = 243389
  * Let's call this IKE, `originalIKE`.
  * Let's call this serial `originalIKESerial`.

### Timeline

* 16:07:00:
  * `originalIKE` replaces a previous IKE via an IKE SA rekey.
* 16:07:52:
  * `crashChild` state object is created.
* 16:07:58:
  * `crashChild.st_accepted_esp_or_ah_proposal` is set to a non-`NULL` value.
     * IKE associated with `crashChild` is `originalIKE`.
  *  `crashChild` encounters `STF_FATAL`.
     * NOTE: This is 1 minute before the crash.
     * The `STF_FATAL` is due to a Netlink add SA error for `unconfined_t` label.
        * NOTE: `unconfined_t` was used many times priorly without any issues.
     * `crashChild` is in state `REKEY_CHILD_I1` when it encounters `STF_FATAL`.
* 16:08:05:
  * `crashIKE` replaces `originalIKE` with a IKE SA rekey:
```
"<TunnelName>" #243550 (crashIKESerial): initiating rekey to replace IKE SA #243389 (originalIKE)
"<TunnelName>" #243550: sent CREATE_CHILD_SA request to rekey IKE SA
"<TunnelName>" #243550: rekeyed #243389 STATE_V2_REKEY_IKE_I1 and expire it remaining life 54.976025s
"<TunnelName>" #243550: established IKE SA {cipher=AES_CBC_256 integ=HMAC_SHA1_96 prf=HMAC_SHA1 group=MODP2048}
```
* 16:08:58:
  * `crashChild` sends a `CREATE_CHILD_SA` request to rekey IPsec SA.
    * `crashChild` is in state `REKEY_CHILD_I1`.
  * The crash happens.
     * Let's call this `crashTime`.
  * `crashTime` is 1 minute after the above `STF_FATAL`.
    * `salifetime` is 1 minute.
    * `ikelifetime` is 2 minutes.

### Crash Backtrace

```
#2  0x00005622b892a7d9 in llog_passert (logger=0x5622b8bf0460 <failsafe_logger>, where=where@entry=0x5622b8be6910 <here>, fmt=fmt@entry=0x5622b898e190 "%s")
    at lib/libswan/passert.c:34
#3  0x00005622b88fc1b9 in ikev2_process_sa_payload (what=what@entry=0x5622b895f6c8 "CREATE_CHILD_SA responder matching remote ESP/AH proposals", sa_payload=sa_payload@entry=0x5622badbbe20, 
    expect_ike=expect_ike@entry=false, expect_spi=expect_spi@entry=true, expect_accepted=expect_accepted@entry=true, opportunistic=<optimized out>, chosen_proposal=0x5622bad64be0, 
    local_proposals=0x5622babe3e68, logger=0x5622bae41c08) at programs/pluto/ikev2_spdb_struct.c:1138
#4  0x00005622b88f7c5c in process_v2_childs_sa_payload (what=what@entry=0x5622b895f6c8 "CREATE_CHILD_SA responder matching remote ESP/AH proposals", ike=ike@entry=0x5622bae4f918, 
    child=child@entry=0x5622bad64648, md=md@entry=0x5622badbbad8, expect_accepted_proposal=expect_accepted_proposal@entry=true)
    at programs/pluto/ikev2_child.c:696
#5  0x00005622b887f2a3 in process_v2_CREATE_CHILD_SA_child_response (ike=0x5622bae4f918, larval_child=0x5622bad64648, response_md=0x5622badbbad8)
    at programs/pluto/ikev2_create_child_sa.c:1018
#6  0x00005622b88f357c in v2_dispatch (ike=ike@entry=0x5622bae4f918, st=st@entry=0x5622bad64648, md=md@entry=0x5622badbbad8, svm=svm@entry=0x5622b8c00fd8 <v2_state_transition_table+1976>)
    at programs/pluto/ikev2.c:2149
#7  0x00005622b88f3788 in ikev2_process_state_packet (ike=ike@entry=0x5622bae4f918, st=st@entry=0x5622bad64648, md=md@entry=0x5622badbbad8)
    at programs/pluto/ikev2.c:2104
#8  0x00005622b88f4167 in ike_process_packet (ike=0x5622bae4f918, md=0x5622badbbad8) at programs/pluto/ikev2.c:1602
#9  ikev2_process_packet (md=md@entry=0x5622badbbad8) at programs/pluto/ikev2.c:1498
#10 0x00005622b890efaa in process_packet (mdp=mdp@entry=0x7ffeeb32c9f8) at programs/pluto/demux.c:241
#11 0x00005622b890f656 in process_md (mdp=0x7ffeeb32c9f8) at programs/pluto/demux.c:348
#12 process_iface_packet (fd=<optimized out>, event=<optimized out>, ifp_arg=0x0) at programs/pluto/demux.c:348
#13 0x00007f2f91850ff1 in event_persist_closure (ev=<optimized out>, base=0x5622baaf4bd8) at event.c:1580
#14 event_process_active_single_queue (base=base@entry=0x5622baaf4bd8, activeq=0x5622bab963f8, max_to_process=max_to_process@entry=2147483647, endtime=endtime@entry=0x0) at event.c:1639
#15 0x00007f2f91851787 in event_process_active (base=0x5622baaf4bd8) at event.c:1738
```

### Logs

These are the log messages relevant to `crashChild` immediately preceding the crash:

```
16:08:58: found #243520 (crashChildSerial) for (find_v2_sa_by_initiator_wip() +1012 programs/pluto/ikev2.c)
16:08:58: found #243520 for (find_v2_sa_by_initiator_wip() +1012 programs/pluto/ikev2.c)
16:08:58: #243520 is idle
16:08:58: #243520 idle
16:08:58: #243520 in state REKEY_CHILD_I1: sent CREATE_CHILD_SA request to rekey IPsec SA
16:08:58: #243550 updating local interface from 192.168.202.11:500 to 192.168.202.11:500 using md->iface (update_ike_endpoints() +2479 programs/pluto/state.c)
16:08:58: forcing ST #243520 to CHILD #243550.#243520 in FSM processor
16:08:58: calling processor process rekey Child SA response (CREATE_CHILD_SA)
16:08:58: using existing local ESP/AH proposals for <TunnelName> (CREATE_CHILD_SA responder matching remote ESP/AH proposals): 1:ESP=AES_CBC_256-HMAC_SHA1_96-MODP2048-DISABLED
16:08:58: comparing remote proposals against CREATE_CHILD_SA responder matching remote ESP/AH proposals 1 local proposals
16:08:58:ABORT: ASSERTION FAILED: *chosen_proposal == ((void *)0) (ikev2_process_sa_payload() +1138 programs/pluto/ikev2_spdb_struct.c)
16:08:58 libreswan1.localdomain systemd[1]: ipsec.service: Main process exited, code=killed, status=6/ABRT
16:08:58 libreswan1.localdomain ip[3820048]: RTNETLINK answers: Permission denied
16:08:58 libreswan1.localdomain systemd[1]: ipsec.service: Control process exited, code=exited status=2
16:08:58 libreswan1.localdomain systemd[1]: ipsec.service: Failed with result 'signal'.
```

I have uploaded 3 log files:

* [crashChild.log](https://github.com/libreswan/libreswan/files/7095963/crashChild.log): Contains all log entries for just `crashChild`.
* [crashChild-crashIKE.log](https://github.com/libreswan/libreswan/files/7095965/crashChild-crashIKE.log): Contains all log entries for `crashChild` + `crashIKE`.
* [crashChild-crashIKE-originalIKE.log](https://github.com/libreswan/libreswan/files/7095967/crashChild-crashIKE-originalIKE.log): Contains all log entries for `crashChild` + `crashIKE` + `originalIKE`.

### Hypothesis

The assert fires because `crashChild` did not get deleted/cleaned up after encountering `STF_FATAL`.

### Questions

* Why didn't the `crashChild` SA not get cleaned up shortly after encountering `STF_FATAL`?

**Comment by cagney** [comment]
The fatal error should have wiped the entire IKE family.  Try reverting 68fb298d059854253e8267680aeee1ee1f3158a3

**Comment by kavinda-forcepoint** [comment]
Is reverting 68fb298 going to bring back the unexpected state deletion issue in https://github.com/libreswan/libreswan/issues/441?

**Comment by cagney** [comment]
No.  In v4.5, the IKE SA isn't deleted when the first child fails because TS was rejected.
Here we've two different problems:
- a fatal error should have deleted the IKE family
- more broadly, what to do when the kernel I/O fails; the code considers it to be fatal, should it instead just fail that child and stumble on
Any ideas as to why the kernel i/o didn't work?

**Comment by cagney** [comment]
this is fixed! see #1609
